modules/core/perf/perf_math.cpp

```diff
@@ -36,6 +36,238 @@ PERF_TEST_P(VectorLength, phase64f, testing::Values(128, 1000, 128*1024, 512*102
+    Mat s(rows, cols, mtype, Scalar(0));
+
+    std::vector<FType> singVals(rank);
+    rng.fill(singVals, RNG::UNIFORM, Scalar(0), Scalar(10));
+    std::set<FType> setSingVals(singVals.begin(), singVals.end());
+    auto singIter = setSingVals.rbegin();
+    for (int i = 0; i < rank; i++)
+    {
+        s.at<FType>(i, i) = *singIter++;
+    }
+
+    if (symmetrical)
+        return u * s * u.t();
+    else
+        return u * s * v.t();
+}
+
+Mat buildRandomMat(int rows, int cols, int mtype, RNG& rng, int rank, bool symmetrical)
+{
+    if (mtype == CV_32F)
+    {
+        return buildRandomMat<float>(rows, cols, rng, rank, symmetrical);
+    }
+    else // if (mtype == CV_64F)
+    {
+        return buildRandomMat<double>(rows, cols, rng, rank, symmetrical);
+    }
+}
+
+CV_ENUM(SolveDecompEnum, DECOMP_LU, DECOMP_SVD, DECOMP_EIG, DECOMP_CHOLESKY, DECOMP_QR)
+
+enum RankMatrixOptions
+{
+    RANK_HALF, RANK_MINUS_1, RANK_FULL
+};
+
+CV_ENUM(RankEnum, RANK_HALF, RANK_MINUS_1, RANK_FULL)
+
+enum SolutionsOptions
+{
+    NO_SOLUTIONS, ONE_SOLUTION, MANY_SOLUTIONS
+};
+
+CV_ENUM(SolutionsEnum, NO_SOLUTIONS, ONE_SOLUTION, MANY_SOLUTIONS)
+
+typedef perf::TestBaseWithParam<std::tuple<int, RankEnum, MatDepth, SolveDecompEnum, bool, SolutionsEnum>> SolveTest;
+
+PERF_TEST_P(SolveTest, randomMat, ::testing::Combine(
+    ::testing::Values(31, 64, 100),
+    ::testing::Values(RANK_HALF, RANK_MINUS_1, RANK_FULL),
+    ::testing::Values(CV_32F, CV_64F),
+    ::testing::Values(DECOMP_LU, DECOMP_SVD, DECOMP_EIG, DECOMP_CHOLESKY, DECOMP_QR),
+    ::testing::Bool(), // normal
+    ::testing::Values(NO_SOLUTIONS, ONE_SOLUTION, MANY_SOLUTIONS)
+    ))
+{
+    auto t = GetParam();
+    int size       = std::get<0>(t);
+    auto rankEnum  = std::get<1>(t);
+    int mtype      = std::get<2>(t);
+    int method     = std::get<3>(t);
+    bool normal    = std::get<4>(t);
+    auto solutions = std::get<5>(t);
+
+    bool symmetrical = (method == DECOMP_CHOLESKY || method == DECOMP_LU);
+
+    if (normal)
+    {
+        method |= DECOMP_NORMAL;
+    }
+
+    int rank = size;
+    switch (rankEnum)
+    {
+    case RANK_HALF:    rank /= 2; break;
+    case RANK_MINUS_1: rank -= 1; break;
+    default: break;
+    }
+
+    RNG& rng = theRNG();
+    while (next())
+    {
+        Mat A = buildRandomMat(size, size, mtype, rng, rank, symmetrical);
+        Mat x(size, 1, mtype);
+        Mat b(size, 1, mtype);
+
+        switch (solutions)
+        {
+        // no solutions, let's make b random
+        case NO_SOLUTIONS: rng.fill(b, RNG::UNIFORM, Scalar(-1), Scalar(1)); break;
+        // exactly 1 solution, let's combine b from A and x
+        case ONE_SOLUTION:
+        {
+            rng.fill(x, RNG::UNIFORM, Scalar(-10), Scalar(10));
+            b = A * x;
+        }
+        break;
+        // infinitely many solutions, let's make b zero
+        default: b = 0; break;
+        }
+
+        startTimer();
+        cv::solve(A, b, x, method);
```

Comment:
@opencv-alalek @vpisarev What do you do about the proposed test? It runs solver with different input on each iteration.

Reply:
Input test data must be fixed between samples measurement.
Input test data must be prepared outside of the main loop.